### ISSUE TITLE
writeNifti produces unreasonably large files and takes way to long

### ISSUE THREAD

**Issue by jadenecke** [jonclayden/RNifti#28]
Sorry for the provocative title, so far I enjoyed your package very much, appreciate the work. However, I am currently working on 4D (fMRI) images and I couldn't help but notice that the writeNifti function is terribly slow. So far with 3D images I haven't noticed anything but with 4D it is way out of hand. Please compare:
```
RNifti::writeNifti(array(runif(5000), dim = c(100, 100, 100, 200)), "test.nii.gz")
oro.nifti::writeNIfTI(array(runif(5000), dim = c(100, 100, 100, 200)), "test2", compression = 5)
```
At first, I thought the code crashed, but no, it just takes a couple of minutes and produces a multi-gigabite sized file. I would have tried to look into it myself but I don't understand C :(

Thanks for you help.

**Comment by jonclayden** [comment]
Hi. Could you provide some platform context, please, such as the output from `sessionInfo()`?

**Comment by jadenecke** [comment]
R version 4.2.1 (2022-06-23)
Platform: x86_64-apple-darwin17.0 (64-bit)
Running under: macOS Mojave 10.14.6

Matrix products: default
BLAS:   /System/Library/Frameworks/Accelerate.framework/Versions/A/Frameworks/vecLib.framework/Versions/A/libBLAS.dylib
LAPACK: /Library/Frameworks/R.framework/Versions/4.2/Resources/lib/libRlapack.dylib

locale:
[1] en_US.UTF-8/en_US.UTF-8/en_US.UTF-8/C/en_US.UTF-8/en_US.UTF-8

attached base packages:
[1] parallel  stats     graphics  grDevices utils     datasets  methods   base     

other attached packages:
[1] tictoc_1.1      pbmcapply_1.5.1 psych_2.2.9     optparse_1.7.3  RNifti_1.4.3   

loaded via a namespace (and not attached):
 [1] Rcpp_1.0.9         cellranger_1.1.0   pillar_1.8.1       compiler_4.2.1     bitops_1.0-7       parameters_0.20.1 
 [7] tools_4.2.1        digest_0.6.31      corrplot_0.92      jsonlite_1.8.4     nlme_3.1-161       evaluate_0.19     
[13] lifecycle_1.0.3    tibble_3.1.8       lattice_0.20-45    pkgconfig_2.0.3    rlang_1.0.6        oro.nifti_0.11.4  
[19] cli_3.6.0          rstudioapi_0.14    yaml_2.3.6         mvtnorm_1.1-3      xfun_0.36          fastmap_1.1.0     
[25] coda_0.19-4        bayestestR_0.13.0  knitr_1.41         tinylabels_0.2.3   vctrs_0.5.1        getopt_1.20.3     
[31] grid_4.2.1         glue_1.6.2         fansi_1.0.3        readxl_1.4.1       rmarkdown_2.19     papaja_0.1.1      
[37] magrittr_2.0.3     splines_4.2.1      htmltools_0.5.4    emmeans_1.8.3      abind_1.4-5        datawizard_0.6.5  
[43] insight_0.18.8     mnormt_2.1.1       effectsize_0.8.2   xtable_1.8-4       utf8_1.2.2         estimability_1.4.1

**Comment by jonclayden** [comment]
Thanks. So, I think the short answer is that `RNifti` isn't the bottleneck here, `zlib` is. If you `gunzip` the resulting files and then `gzip` them again from the command-line, the performance is no better, either in compression ratio or speed:

```
$ time gzip test.nii

________________________________________________________
Executed in  109.82 secs    fish           external
   usr time  107.22 secs    0.09 millis  107.22 secs
   sys time    0.71 secs    1.19 millis    0.71 secs

$ time gzip test2.nii

________________________________________________________
Executed in    3.26 secs    fish           external
   usr time    3.12 secs    0.10 millis    3.12 secs
   sys time    0.10 secs    1.38 millis    0.10 secs
```

My stack traces confirm the same thing: when called from `RNifti`, almost all the time is taken by DEFLATE, the `zlib` compression function.

So, why does `oro.nifti` seem to do so much better? Well, that's where things start to get interesting. The unzipped version of `test2.nii` is half the size of `test.nii`, because `oro.nifti` is actually choosing the single-precision floating-point datatype for the image, whereas `RNifti` defaults to double-precision because that's R's native type. You can get very similar results (speed and size wise) if you ask `RNifti` to follow suit:

```r
RNifti::writeNifti(array(runif(5000), dim = c(100, 100, 100, 200)), "test3.nii.gz", datatype="float")
```

What's going on? Well, random data (from `runif()`) is worst-case for compression algorithms, because entropy is maximal, so the 5000 generated values are essentially incompressible. Their single-precision size is 5000\*4 = 20,000 bytes, and their double-precision size is 5000\*8 = 40,000 bytes, or 39 KiB. But DEFLATE [searches for repetitions over a *32 KiB* sliding window](https://en.wikipedia.org/wiki/Deflate#Duplicate_string_elimination), meaning that it won't look back far enough to spot the repetitions and exploit them. So it does not much better than no compression in terms of the final file size—the uncompressed size is 8\*100\*100\*100\*200 bytes, or 1.5 GiB—but wastes a lot of time trying. This doesn't apply to the single-precision case, because the repeating block is smaller than the window. Perhaps @madler will correct my analysis if it's wrong, but it seems to line up with the information I have found (including from perusal of the source code).

In short, whether you intended to or not, you seem to have created a pathological case of an image, with hard-to-compress data that repeats just rarely enough that the default `zlib` parameters will not result in back-references and therefore effective compression. I don't think there's any realistic way around this, because [`zlib` doesn't support a larger window than 2^15 bytes, i.e., 32 KiB](https://github.com/madler/zlib/blob/04f42ceca40f73e2978b50e93806c2a18c1281fc/zlib.h#L552). Hopefully this kind of data won't pop up much in practice—but if it does, you may want to store it uncompressed.

**Comment by madler** [comment]
Yes, that analysis is correct.

While that example is somewhat contrived, I can imagine that there might be real-world cases where the data is not compressible. I would recommend using level 1 for the compression, so that you're not wasting much time trying to compress something that can't be, but still get most of the benefit when it can be. Since this is image data and not text, there is little benefit to searching harder for string matches, which is what the compression level controls.

**Comment by jonclayden** [comment]
Many thanks for confirming, Mark.

**Comment by jadenecke** [comment]
Thank you very much for your detailed analysis. That already explains a lot. My use-case is a mean fMRI image, it goes from 32-bit 700Mb uncompressed to 170Mb compressed. Both functions now produce the same file size, however, there are still some differences which I cant explain. I usually work on a NFS mount. Storing files locally, RNifti is faster than oro.nifti (compressed files). For storing files on the network drive, there is no difference between both algorithms when saving uncompressed, however when storing compressed, RNifti takes twice as long as oro.nifti. My only naive guess would be that oro.nifti compresses in memory and than writes to disk, with it being somehow different for RNifti?

**Comment by jonclayden** [comment]
The two packages take quite a different route to the compression functions: `oro.nifti` uses R's `writeBin` function and thence, via the connection abstraction, R's own interface to `zlib`, whereas `RNifti` uses the [NIfTI reference C library](https://github.com/NIFTI-Imaging/nifti_clib) which interfaces with `zlib` via a thin layer called `znzlib`. There are certain differences in the decisions made along the way in the two cases.

If I'm reading the source correctly, one of those differences is in the size of the output buffer used: `zlib` [defaults to 8 KiB](https://github.com/madler/zlib/blob/04f42ceca40f73e2978b50e93806c2a18c1281fc/gzguts.h#L156), whereas [R uses 16 KiB](https://github.com/wch/r-source/blob/46f31a055b72a2187af58c7f78bed3fbceb11b6a/src/main/gzio.h#L56). I suspect, therefore, that `oro.nifti`'s implicitly larger buffer is leading to more efficient network requests, with a data payload around twice the size, and that results in the better write performance that you see—basically in line with your intuition.

Unlike the sliding window width, but like the compression level, `zlib`'s API provides a way to change the buffer size (just) after the file is opened, so `RNifti` could in principle provide a means to tweak this. But this is pretty low-level stuff, so I'm not sure many people would use such a facility!

**Comment by jadenecke** [comment]
Thank you very much for your explanations, I definitively learned a lot!

**Comment by jonclayden** [comment]
Me too! 🙂

**Comment by jonclayden** [comment]
@jadenecke In the latest development version of the package, I've added a mechanism to provide control over the buffer size. If you have a chance to try it, I'd be interested to know whether setting

```r
options(RNifti.gzBufferSize=16384)
```

affects the speed of the network writes that you were seeing...

**Comment by jadenecke** [comment]
I am afraid that for my setup there is no difference before and after setting the buffer size. Would there be a way to verify that its actually writing with a different buffer size?

**Comment by jadenecke** [comment]
Sorry for the late response, I have now got around to testing it. Unfortunately it does not seem to make a difference in my case. However I think there is something wrong in general. It should not take this much time in general and I assume its a different issue altogether. I checked that the protocol is NFS3 and the max. buffer size is 32Kb. 

``` R
> system.time(oro.nifti::writeNIfTI(m, "oroNifti", compression = 5))
   user  system elapsed 
  0.873   0.098  16.849 
> options(RNifti.gzBufferSize=8192)
> system.time(RNifti::writeNifti(m, "RNifit_normalBuffer.nii.gz", datatype="float"))
** znzbuffer: setting buffer size to 8192
   user  system elapsed 
  0.816   0.172  36.338 
> options(RNifti.gzBufferSize=16384)
> system.time(RNifti::writeNifti(m, "RNifit_largerBuffer.nii.gz", datatype="float"))
** znzbuffer: setting buffer size to 16384
   user  system elapsed 
  0.823   0.128  41.008 
```

**Comment by jonclayden** [comment]
Hmm – thanks for trying. Are these results reproducible, i.e., if you run the write two or three times, does it take roughly the same time each time?

**Comment by jadenecke** [comment]
roughly, yes. There is some variations, sometimes the larger buffer-size leads by a couple of seconds sometimes vice versa. However oro.nifti is always approx. twice as fast.

**Comment by jonclayden** [comment]
Thanks. I can't discern any other particular reason that might be behind this, unfortunately. I don't have an NFS share readily available to try with, but when writing to an SMB share with your commands above I actually find that the `RNifti` calls take less than half the time of the `oro.nifti` one (with a small additional advantage with the larger buffer). Given the relative complexity of network writes I think this is ultimately out of scope, but if you can pin anything down that you think originates in the package code, do open a new issue. All the best.

**Comment by jadenecke** [comment]
Absolutely understandable, thank you very much for you efforts! If I ever figure anything out (unlikely though), I will let you know.
